processing/src/main/java/io/druid/segment/QueryableIndexIndexableAdapter.java

```diff
@@ -247,22 +217,35 @@ public Rowboat next()
               throw new NoSuchElementException();
             }
 
-            final int[][] dims = new int[dictionaryEncodedColumns.length][];
+            final Comparable[][] dims = new Comparable[columns.length][];
             int dimIndex = 0;
-            for (final DictionaryEncodedColumn dict : dictionaryEncodedColumns) {
-              final IndexedInts dimVals;
-              if (dict.hasMultipleValues()) {
-                dimVals = dict.getMultiValueRow(currRow);
-              } else {
-                dimVals = new ArrayBasedIndexedInts(new int[]{dict.getSingleValueRow(currRow)});
-              }
+            for (final Closeable column : columns) {
+              if (column instanceof DictionaryEncodedColumn) {
+                DictionaryEncodedColumn dict = (DictionaryEncodedColumn) column;
+                final IndexedInts dimVals;
+                if (dict.hasMultipleValues()) {
+                  dimVals = dict.getMultiValueRow(currRow);
+                } else {
+                  dimVals = new ArrayBasedIndexedInts(new int[]{dict.getSingleValueRow(currRow)});
+                }
 
-              int[] theVals = new int[dimVals.size()];
-              for (int j = 0; j < theVals.length; ++j) {
-                theVals[j] = dimVals.get(j);
-              }
+                Integer[] theVals = new Integer[dimVals.size()];
+                for (int j = 0; j < theVals.length; ++j) {
+                  theVals[j] = dimVals.get(j);
+                }
 
-              dims[dimIndex++] = theVals;
+                dims[dimIndex++] = theVals;
+              } else if (column instanceof GenericColumn) {
+                GenericColumn genericCol = (GenericColumn) column;
+                switch(genericCol.getType()) {
+                  case LONG:
```

Comment:
as i read through the code, im a little worried of all the places that require switch statements such as this. It'll be easy to miss cases and cause bugs as we need new types. I wonder if we can commonalize this logic somehow